hypervisor/common/hypercall.c

```diff
@@ -1216,16 +1219,31 @@ int32_t hcall_destroy_vdev(struct acrn_vm *vm, struct acrn_vm *target_vm, __unus
 {
 	int32_t ret = -EINVAL;
 	struct acrn_emul_dev dev;
+	struct pci_vdev *vdev;
 	struct emul_dev_ops *op;
+	union pci_bdf bdf;
 
 	/* We should only destroy a device to a post-launched VM at creating or pausing time for safety, not runtime or other cases*/
 	if (is_created_vm(target_vm) || is_paused_vm(target_vm)) {
 		if (copy_from_gpa(vm, &dev, param2, sizeof(dev)) == 0) {
-		op = find_emul_dev_ops(&dev);
-		if ((op != NULL) && (op->destroy != NULL)) {
-			ret = op->destroy(target_vm, &dev);
+			op = find_emul_dev_ops(&dev);
+			if (op != NULL) {
+				bdf.value = (uint16_t) dev.slot;
+				vdev = pci_find_vdev(&vm->vpci, bdf);
+				if (vdev != NULL) {
+					vdev->pci_dev_config->vbdf.value = UNASSIGNED_VBDF;
+					ret = 0;
+				} else {
+					pr_warn("%s, failed to destroy emulated device %x:%x.%x\n",
+						__func__, bdf.bits.b, bdf.bits.d, bdf.bits.f);
+					ret = -EINVAL;
+				}
+
+				if (op->destroy != NULL) {
+					ret = op->destroy(vdev);
```

Comment:
The previous "ret" got in Line 1235 was overwritten then

Reply:
Or if vdev == NULL, do we need to do this?

Reply:
Yes, that is a bug